src/citrine/builders/auto_configure.py

```diff
@@ -0,0 +1,183 @@
+from citrine.resources.material_run import MaterialRun
+from citrine.resources.project import Project
+from citrine.resources.table_config import TableBuildAlgorithm
+
+
+def extract_descriptor_keys(
+        *,
+        project: Project,
+        material: Union[MaterialRun, LinkByUID, str, UUID],
+        mode: str = 'SIMPLE',
+        full_history: bool = False
+) -> Dict:
+    """[ALPHA] Extract names of attributes from a provided material history.
+
+    Given a material (or link representation), extracts the names from
+    property, condition, and parameter templates
+    that appear within process and measurement runs
+    throughout the material's history.
+
+    The obtained descriptor keys can be used when constructing
+    objectives for a scoring function in a design workflow.
+
+    Parameters
+    ----------
+    project: Project
+        Project to use when accessing the Citrine Platform
+    material: Union[MaterialRun, LinkByUID, str, UUID]
+        Sample material to extract descriptor keys
+    mode: str
+        Option in {'SIMPLE', 'FORMULATION'} that controls formatting descriptor keys
+        Default: 'SIMPLE'
+    full_history: bool
+        Whether to extract descriptor keys from the full material history,
+        or only the provided (terminal) material
+        Default: True
+
+    Returns
+    -------
+    dict
+        Dictionary of keys of the form
+        {'properties': [...], 'conditions': [...], 'parameters': []}
+
+    """
+    if mode not in {"FORMULATION", "SIMPLE"}:
+        msg = "Called with mode: {}.  Expected 'SIMPLE' or 'FORMULATION'.".format(mode)
+        raise ValueError(msg)
+
+    link = _make_link_by_uid(material)
+    history = project.material_runs.get_history(scope=link.scope, id=link.id)
+
+    if full_history:
+        search_history = flatten(history)
+    else:
+        # Limit the search to contain the terminal material/process/measurement
+        search_history = [history.spec.template, history.process.template]
+        search_history.extend([msr.template for msr in history.measurements])
+
+    keys = {'properties': [], 'conditions': [], 'parameters': []}
+    for obj in search_history:
+        parent_name = f'{obj.name}~' if mode == 'SIMPLE' else ''
+        # Search all attribute template types
+        if isinstance(obj, HasPropertyTemplates):
+            for property in obj.properties:
+                keys['properties'].append(parent_name + property[0].name)
+        if isinstance(obj, HasConditionTemplates):
+            for condition in obj.conditions:
+                keys['conditions'].append(parent_name + condition[0].name)
+        if isinstance(obj, HasParameterTemplates):
+            for parameter in obj.parameters:
+                keys['parameters'].append(parent_name + parameter[0].name)
+    return keys
+
+
+def auto_configure_material(
+        *,
+        project: Project,
+        material: Union[MaterialRun, LinkByUID, str, UUID],
+        score: Score,
```

Comment:
I think just doing an objective could be even simpler.  Or you could have both as options and just make sure either the score or the objective is given.

Reply:
If specifying an objective (or list of objectives), how should we include baseline information? My only thought would be to read the table, determine appropriate min/max of training data columns, and set the appropriate baseline given which type of objective.

Although at that point, maybe it is safer to have the user pass a pre-constructed Score where they can control those parameters?